bee-network/bee-autopeering/src/discovery/manager.rs

```diff
@@ -0,0 +1,1000 @@
+
+        master_peers.write().insert(*peer.peer_id());
+
+        // Also add it as a regular peer.
+        if let Some(peer_id) = add_peer::<false>(peer, local, active_peers, replacements) {
+            log::debug!("Added {}.", peer_id);
+            num_added += 1;
+        }
+    }
+
+    log::debug!("Added {} entry node/s.", num_added);
+}
+
+/// Attempts to add a new peer to a peer list (preferably as active).
+/// If the peer is added inbound, i.e. , the "last verification timestamp" is added.
+pub(crate) fn add_peer<const ON_REQUEST: bool>(
+    peer: Peer,
+    local: &Local,
+    active_peers: &ActivePeersList,
+    replacements: &ReplacementList,
+) -> Option<PeerId> {
+    // Only add new peers.
+    if peer::is_known(peer.peer_id(), local, active_peers, replacements) {
+        None
+    } else {
+        let peer_id = *peer.peer_id();
+        // First try to add it to the active peer list. If that list is full, add it to the replacement list.
+        if !active_peers.read().is_full() {
+            let active_peer = if ON_REQUEST {
+                let mut active = ActivePeer::from(peer);
+                active.metrics_mut().set_last_verif_request_timestamp();
+                active
+            } else {
+                ActivePeer::from(peer)
+            };
+            if active_peers.write().insert(active_peer) {
+                Some(peer_id)
+            } else {
+                None
+            }
+        } else if replacements.write().insert(peer) {
+            Some(peer_id)
+        } else {
+            None
+        }
+    }
+}
+
+// Note: this function is dead-lock danger zone!
+/// Deletes a peer from the active peerlist if it's not a master peer, and replaces it by a peer
+/// from the replacement list.
+pub(crate) fn remove_peer_from_active_list(
+    peer_id: &PeerId,
+    master_peers: &MasterPeersList,
+    active_peers: &ActivePeersList,
+    replacements: &ReplacementList,
+    event_tx: &EventTx,
+) {
+    let mut active_peers = active_peers.write();
+
+    if let Some(mut removed_peer) = active_peers.remove(peer_id) {
+        // hive.go: master peers are never removed
+        if master_peers.read().contains(removed_peer.peer_id()) {
+            // hive.go: reset verifiedCount and re-add them
+            removed_peer.metrics_mut().reset_verified_count();
+            active_peers.insert(removed_peer);
+        } else {
+            // TODO: why is the event only triggered for verified peers?
+            // ```go
+            // if mp.verifiedCount.Load() > 0 {
+            //     m.events.PeerDeleted.Trigger(&DeletedEvent{Peer: unwrapPeer(mp)})
+            // }
+            // ```
+            if removed_peer.metrics().verified_count() > 0 {
+                event_tx
+                    .send(Event::PeerDeleted { peer_id: *peer_id })
+                    .expect("error sending `PeerDeleted` event");
+            }
+
+            // ```go
```

Comment:
:carlos:

Reply:
Just my markdown habit to add the language specifier.